src/func.rs

```diff
@@ -144,15 +145,140 @@ impl FuncInstance {
 		check_function_args(func.signature(), &args).map_err(|_| TrapKind::UnexpectedSignature)?;
 		match *func.as_internal() {
 			FuncInstanceInternal::Internal { .. } => {
-				let mut interpreter = Interpreter::new(externals);
-				interpreter.start_execution(func, args)
+				let mut interpreter = Interpreter::new(func, args)?;
+				interpreter.start_execution(externals)
 			}
 			FuncInstanceInternal::Host {
 				ref host_func_index,
 				..
 			} => externals.invoke_index(*host_func_index, args.into()),
 		}
 	}
+
+	/// Invoke the function, get a resumable handle.  This handle can then be used to actually start the execution. If a
+	/// Host trap happens, caller can use `resume_execution` to feed the expected return value back in, and then
+	/// continue the execution.
```

Comment:
I think it is a good idea to add a note that this is an experimental API. Also it might be a good idea to also warn the user that they might not be able to find anything like this API on other execution engines.